File: java/src/test/java/it/giannibombelli/wsc2026/giftcard/api/GiftCardIssuanceApiTest.java
package it.giannibombelli.wsc2026.giftcard.api;

import it.giannibombelli.wsc2026.giftcard.GiftCardModule;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import it.giannibombelli.wsc2026.testsupport.JavalinTestHelper;
import org.junit.jupiter.api.*;
import tools.jackson.databind.ObjectMapper;

import javax.sql.DataSource;
import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;


@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class GiftCardIssuanceApiTest {
    private static final ObjectMapper JSON = new ObjectMapper();
    private final JavalinTestHelper javalin = new JavalinTestHelper();

    @BeforeAll
    void setUp() {
        final DataSource dataSource = DatabaseSetup.initializeFileDb("giftcard", getClass().getSimpleName());
        GiftCardModule module = new GiftCardModule(dataSource, List.of());

        javalin.start(module::configure);
    }

    @AfterAll
    void tearDown() {
        javalin.stop();
    }

    @Nested
    class PostIssuance {
        @Test
        void shouldReturn201() {
            JavalinTestHelper.TestResponse postResponse = javalin.post("/gift-cards");

            assertThat(postResponse.status()).isEqualTo(201);
        }

        @Test
        void shouldReturnLocationHeader() {
            JavalinTestHelper.TestResponse postResponse = javalin.post("/gift-cards");

            String location = postResponse.header("Location");
            assertThat(location).matches("/gift-cards/[0-9a-fA-F-]{36}");
        }

        @Test
        void shouldReturnCreatedCard() {
            JavalinTestHelper.TestResponse postResponse = javalin.post("/gift-cards");

            GiftCardResponse created = JSON.readValue(postResponse.body(), GiftCardResponse.class);
            assertThat(created.id()).isNotNull();
            assertThat(created.balance()).isEqualByComparingTo(BigDecimal.ZERO);
        }
    }

    @Nested
    class GetById {
        @Test
        void shouldReturn200() {
            JavalinTestHelper.TestResponse postResponse = javalin.post("/gift-cards");
            String location = postResponse.header("Location");

            JavalinTestHelper.TestResponse getResponse = javalin.get(location);

            assertThat(getResponse.status()).isEqualTo(200);
        }

        @Test
        void shouldReturnCard() {
            JavalinTestHelper.TestResponse postResponse = javalin.post("/gift-cards");
            GiftCardResponse created = JSON.readValue(postResponse.body(), GiftCardResponse.class);
            String location = postResponse.header("Location");

            JavalinTestHelper.TestResponse getResponse = javalin.get(location);

            GiftCardResponse readModel = JSON.readValue(getResponse.body(), GiftCardResponse.class);
            assertThat(readModel.id()).isEqualTo(created.id());
            assertThat(readModel.balance()).isEqualByComparingTo(BigDecimal.ZERO);
        }

        @Test
        void shouldReturn404() {
            JavalinTestHelper.TestResponse response = javalin.get("/gift-cards/" + UUID.randomUUID());
            assertThat(response.status()).isEqualTo(404);
        }

        @Test
        void shouldReturn400() {
            JavalinTestHelper.TestResponse response = javalin.get("/gift-cards/invalid-uuid");
            assertThat(response.status()).isEqualTo(400);
            assertThat(response.body()).isEqualTo("Invalid gift card id format");
        }
    }
}

File: typescript/tests/giftcard/api/giftCardIssuanceApi.test.ts
import { beforeAll, describe, expect, it } from 'vitest';
import express from 'express';
import request from 'supertest';
import { GiftCardModule } from '@/giftcard/module.js';
import { DatabaseSetup } from '../../testsupport/databaseSetup.js';

describe('GiftCardIssuanceApi', () => {
  const app = express();
  let module: GiftCardModule;

  beforeAll(() => {
    const database = DatabaseSetup.initializeFileDb('giftcard', 'GiftCardIssuanceApiTest');
    module = new GiftCardModule(database, []);
    module.configure(app);
  });

  describe('post issuance', () => {
    it('should return 201', async () => {
      const response = await request(app).post('/gift-cards');

      expect(response.status).toEqual(201);
    });

    it('should return location header', async () => {
      const response = await request(app).post('/gift-cards');

      expect(response.headers['location']).toMatch(/\/gift-cards\/[0-9a-fA-F-]{36}/);
    });

    it('should return created card', async () => {
      const response = await request(app).post('/gift-cards');

      expect(response.body.id).not.toBeNull();
      expect(response.body.balance).toEqual(0);
    });
  });

  describe('getById', () => {
    it('should return 200', async () => {
      const created = await request(app).post('/gift-cards');
      const location = created.headers['location'] as string;

      const response = await request(app).get(location);

      expect(response.status).toEqual(200);
    });

    it('should return card', async () => {
      const created = await request(app).post('/gift-cards');
      const createdBody = created.body as { id: string; balance: number };
      const location = created.headers['location'] as string;

      const response = await request(app).get(location);

      expect(response.body.id).toEqual(createdBody.id);
      expect(response.body.balance).toEqual(0);
    });

    it('should return 404', async () => {
      const response = await request(app).get(`/gift-cards/${crypto.randomUUID()}`);
      expect(response.status).toEqual(404);
    });

    it('should return 400', async () => {
      const response = await request(app).get('/gift-cards/invalid-uuid');
      expect(response.status).toEqual(400);
      expect(response.text).toEqual('Invalid gift card id format');
    });
  });
});
